enhancements/cloud-integration/tokenized-auth-enablement-operators-on-cloud.md

```diff
@@ -0,0 +1,493 @@
+deployed. Currently, several Red Hat operators provide detailed instructions for how to use the operator on STS-enabled
+clusters. The process is designed to have manual operations performed by the customer. This enhancement seeks to bring common standards to the process of enabling an
+operator to access cloud provider resources, by enabling automation, UX and logic for the process. 
+
+## Motivation
+
+Several important Red Hat operators have normal operating modes where they assure their operands have required cloud 
+resources to perform properly. Currently, when a cluster is installed on a cloud platform an admin is required to 
+follow instructions (for example [1]) to provide STS-based credentials to the operator such that needed resources can be
+accessed by the operand. This enhancement seeks to normalize this process across operators, so users of several of them
+have the same experience and similar steps to perform. We aim to automate and reduce those steps as much as possible to 
+make adding operators (via OLM) simpler on an STS enabled cluster.
+
+### User Stories
+
+* As a cluster admin I know which OLM Operators support tokenized authentication for my cloud.
+* As a cluster admin of a cluster using tokenized cloud auth I know what's required to install and upgrade OLM Operators whenever those operators manage resources that authenticate against my cloud.
+* As an Operator developer, I have a standard framework to define tokenized authentication requirements and consume them, per supported cloud.
+* As an Operator Hub browser I know which operators support tokenized cloud auth and on which clouds.
+* As the Hypershift team, where the only credential mode for clusters/customers is STS (on AWS) , the Red Hat branded
+  Operators that must reach the AWS API, should be enabled to work with STS credentials in a consistent, and automated
+  fashion that allows customer to use those operators as easily as possible, driving the use of layered products.
+
+### Goals
+
+Allow Day-2 Operators to access cloud provider resources as seamlessly as possible when being installed, used and
+updated on STS enabled clusters.
+
+While providing the above goal, allow for multi-tenancy. In this sense multi-tenancy means that an operator may
+provide different cloud provider credentials per operand such that operator users may uniquely access non-shared cloud
+resources.
+
+Operator authors have a way to notify, guide, and assist Day-2 Operator admins in providing the required cloud provider
+credentials matched to their permission needs for install and update.
+
+Ideally, a solution here will work in both HyperShift (STS always) and non-HyperShift clusters.
+
+### Non-Goals
+
+Day 1 operators are not included in this proposal, they are pre-provisioned for STS enabled clusters (at least 
+for HyperShift)
+
+
+
+## Proposal
+
+**Cloud Credential Operator (CCO) changes**: Adds a token-aware mode, while nominally set in "Manual" mode, CCO will
```

Comment:
So this token aware mode comes on automatically in manual mode? How will this affect upgraded clusters?

Reply:
my assumption was that it would only processes credreqs that have an ARN(i.e. STS token credrequests), not generally for all credreqs, though now as i read this i don't see that called out.

Reply:
Can you expand the question a bit more with what you're concerned about? As opposed to adding another mode to CCO the proposal is to make several checks in CCO when in Manual mode to detect and handle time-based token auth (STS on AWS) enabled clusters differently. Not sure I understand how this logic is different for upgrades?

Reply:
the concern is that on upgrade, if the CCO is in manual mode, it could suddenly start processing CredReqs it had previously ignored, which would be surprising and undesirable behavior.

Reply:
what do you mean by process the CredReqs? In manual mode CCO doesn't do anything with the credReqs afaik.

Reply:
can you clarify why an operator upgrade might change CCO's behavior toward the CredReqs created by the operator?

Reply:
> what do you mean by process the CredReqs? In manual mode CCO doesn't do anything with the credReqs afaik.

I had suggested in [comment](https://github.com/openshift/enhancements/pull/1339#discussion_r1170592044) that CCO remain in `Manual` mode for this but only reconcile CredenditlasRequests that have the new fields when the cluster has a `serviceAccountIssuer` + platform=AWS etc indicating that the cluster where CCO is running is STS-capable when in `Manual` mode.